### ISSUE TITLE
Unrecognized associated type bound on another associated type

### ISSUE THREAD

**Issue by kvark** [rust-lang/rust#24159]
``` rust
trait Bar<T> { fn dummy(&self); }
trait Car<T> { fn dummy(&self); }

trait Foo {
    type A;
    type B: Bar<Self::A>;
    type C: Car<Self::A>;

    fn get_b(&self) -> &Self::B;
}

fn test_bar<A, B: Bar<A>>(_: &B) {}

fn test<A, F: Foo<A=A>>(f: &F) {
    test_bar(f.get_b());
}
```

Gives me:

```
<anon>:15:16: 15:23 error: the trait `Bar<A>` is not implemented for the type `<F as Foo>::B` [E0277]
<anon>:15     test_bar(f.get_b());
                         ^~~~~~~
<anon>:15:16: 15:23 error: the trait `Car<A>` is not implemented for the type `<F as Foo>::C` [E0277]
<anon>:15     test_bar(f.get_b());
                         ^~~~~~~
error: aborting due to 2 previous errors
playpen: application terminated with error code 101
```

There is a workaround:

``` rust
fn test<A, B: Bar<A>, C: Car<A>, F: Foo<A=A, B=B, C=C>>(f: &F) {
    test_bar(f.get_b());
}
```

But it's ugly and should not be necessary.

We hit this problem a lot with [gfx::Device](https://github.com/gfx-rs/gfx-rs/blob/68298c260b8d420c9c1cfc8657c87462a84476a6/src/device/mod.rs#L233), and I'd like to see cleaner use of it without explicit `CommandBuffer` bounds that we use as a [workaround](https://github.com/PistonDevelopers/gfx-debug-draw/blob/58e6d6206a0632492723566ce018ce5a85536c35/src/text_renderer.rs#L246).

**Comment by apasel422** [comment]
This may be related to #23481.

**Comment by brson** [comment]
This is the error now:

```


rustc 1.13.0 (2c6933acc 2016-11-07)
error[E0405]: trait `std::marker::PhantomFn` is not in scope
 --> <anon>:1:15
  |
1 | trait Bar<T>: std::marker::PhantomFn<T> { fn dummy(&self); }
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^ `std::marker::PhantomFn` is not in scope
  |
  = help: no candidates by the name of `PhantomFn` found in your project; maybe you misspelled the name or forgot to import an external crate?

error[E0405]: trait `std::marker::PhantomFn` is not in scope
 --> <anon>:2:15
  |
2 | trait Car<T>: std::marker::PhantomFn<T> { fn dummy(&self); }
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^ `std::marker::PhantomFn` is not in scope
  |
  = help: no candidates by the name of `PhantomFn` found in your project; maybe you misspelled the name or forgot to import an external crate?

error: main function not found

error: cannot continue compilation due to previous error

```

**Comment by Mark-Simulacrum** [comment]
Edited original post to remove `PhantomFn` bounds, they're no longer needed. Issue still reproduces:

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> test.rs:14:1
   |
14 | / fn test<A, F: Foo<A=A>>(f: &F) {
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`
   |
   = help: consider adding a `where <F as Foo>::B: Bar<A>` bound
   = note: required by `Foo`

error[E0277]: the trait bound `<F as Foo>::C: Car<A>` is not satisfied
  --> test.rs:14:1
   |
14 | / fn test<A, F: Foo<A=A>>(f: &F) {
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Car<A>` is not implemented for `<F as Foo>::C`
   |
   = help: consider adding a `where <F as Foo>::C: Car<A>` bound
   = note: required by `Foo`

error: aborting due to 2 previous errors
```

**Comment by rkarp** [comment]
It looks like a lot of related issues for this have been opened over time.

Duplicates: #37808, #39532, #40093

Probably same underlying cause: #32722, #37883, #38917, #43475

There's also a related discussion on the Rust forums: https://users.rust-lang.org/t/bounds-on-associated-types-enforced-too-soon/12948

**Comment by nikomatsakis** [comment]
Mildly simplified:

```
trait Bar<T> { fn dummy(&self); }

trait Foo {
    type A;
    type B: Bar<Self::A>;

    fn get_b(&self) -> &Self::B;
}

fn test_bar<A, B: Bar<A>>(_: &B) {}

fn test<A, F: Foo<A=A>>(f: &F) {
    test_bar(f.get_b());
}

fn main() { }
```

So, probably the fix is "just" that we need to do some normalization somewhere or other. My *preference* is to fix this by moving over to lazy normalization and chalk, which will do away with these sorts of shortcomings for good. But it's probably worth looking to see if this can be fixed a touch more rapidly by adding a normalization in the right place. I'll try to do that in a bit and mentor/open a fix if it seems reasonable.

**Comment by malbarbo** [comment]
Any progress here @nikomatsakis?

**Comment by ZerothLaw** [comment]
Looks like the issue is known and mentioned at this line: https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/compare_method.rs#L252

This indicates this may be the area to check in when working on this? Still working on understanding a lot of this code, so I'm not sure.

**Comment by daboross** [comment]
Ran into this as well. Here's my minimal test case:

```
trait Bar<T> {}
trait Caz {
    type A;
    type B: Bar<Self::A>;
}

fn test<T, U>(_: T) where T: Caz, U: Caz<A = T::A> {}
```

```text
error[E0277]: the trait bound `<U as Caz>::B: Bar<<T as Caz>::A>` is not satisfied
 --> src/main.rs:7:1
  |
7 | fn test<T, U>(_: T) where T: Caz, U: Caz<A = T::A> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Bar<<T as Caz>::A>` is not implemented for `<U as Caz>::B`
  |
  = help: consider adding a `where <U as Caz>::B: Bar<<T as Caz>::A>` bound
note: required by `Caz`
 --> src/main.rs:2:1
  |
2 | trait Caz {
  | ^^^^^^^^^
```

**Comment by elidupree** [comment]
I've just run into this issue in one of my own projects, and reduced it to a test case that's even simpler than the last one:

```rust
trait Bar<T> {}
trait Caz {
    type A;
    type B: Bar<Self::A>;
}
fn test<T: Caz<A = ()>>() {}
```
Although it compiles successfully if you add the bizarrely-tautological-looking `where T::B: Bar<T::A>`.

For now, I'm stuck with adding the where clause, but that's putting a LOT of boilerplate in my code (the actual where clause is 71 characters long and appears in dozens of places, including being de-facto required in code that uses my library, not just inside the library itself). Is there any progress on this?

**Comment by estebank** [comment]
[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=90590bd4d71c4c30ca1cd791a1d45870):

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> src/lib.rs:14:1
   |
4  |   trait Foo {
   |   --------- required by `Foo`
...
14 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::B: Bar<A>`
   |  _|
   | |
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`

error[E0277]: the trait bound `<F as Foo>::C: Car<A>` is not satisfied
  --> src/lib.rs:14:1
   |
4  |   trait Foo {
   |   --------- required by `Foo`
...
14 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::C: Car<A>`
   |  _|
   | |
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Car<A>` is not implemented for `<F as Foo>::C`
```

[Applying the suggestions](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e43b5d6a74a5815b625691808420f3a0) makes the code compile.

---

[Niko's minimized version](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=056093e1739899755b3f962da894dc35):

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> src/main.rs:12:1
   |
3  |   trait Foo {
   |   --------- required by `Foo`
...
12 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::B: Bar<A>`
   |  _|
   | |
13 | |     test_bar(f.get_b());
14 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`
```

---

@daboross' [case](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7afcdab09435cf3010f0c985a5aaa0b6):

```
error[E0277]: the trait bound `<U as Caz>::B: Bar<<T as Caz>::A>` is not satisfied
 --> src/lib.rs:7:1
  |
2 | trait Caz {
  | --------- required by `Caz`
...
7 | fn test<T, U>() where T: Caz, U: Caz<A = T::A> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-^^
  | |                                             |
  | |                                             help: consider further restricting the associated type: `, <U as Caz>::B: Bar<<T as Caz>::A>`
  | the trait `Bar<<T as Caz>::A>` is not implemented for `<U as Caz>::B`
```

---

In call cases applying the suggestions fulfill the requirements.

**Comment by dignifiedquire** [comment]
@nikomatsakis any update on this? Is there anything I can do help with this? I'd be happy and try to help where I can, as this is blocking a lot of nice usages of `typenum` for me at the moment.

**Comment by daboross** [comment]
It looks like this compiles under `-Z chalk`, so it might just be waiting on fully stable chalk integration.

If that's the case, then this falls into the same bucket as features like [implied bounds](https://github.com/rust-lang/rust/issues/44491), where there isn't a lot we can do outside of just pushing chalk integration all the way.

**Comment by nikomatsakis** [comment]
I think my [original analysis](https://github.com/rust-lang/rust/issues/24159#issuecomment-372731855 is still correct. I think the normalizing is missing here:

https://github.com/rust-lang/rust/blob/ff4a2533a0720f9cdd86e02eafa3725f07aa7752/src/librustc_trait_selection/traits/select/mod.rs#L1295

After we invoke `instantiate`, we should also be normalizing, probably kind of like this

https://github.com/rust-lang/rust/blob/ff4a2533a0720f9cdd86e02eafa3725f07aa7752/src/librustc_trait_selection/traits/select/mod.rs#L1855-L1863

**Comment by estebank** [comment]
All of [the cases I mentioned](https://github.com/rust-lang/rust/issues/24159#issuecomment-549617767) now compile. I believe this can be closed by adding tests.
